### PR TITLE
(kubernetes-cli) Update to the latest supported versions

### PR DESCRIPTION
**Pull request by rgl**
## Description

Update `kubernetes-cli` to the latest supported versions.

## Motivation and Context

See https://github.com/chocolatey-community/chocolatey-packages/discussions/1901#discussioncomment-3082957

## How Has this Been Tested?

I've tested (un)installing in my machine.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Migrated package (a package has been migrated from another repository)

## Checklist:

- [ ] My code follows the code style of this repository.
- [ ] My change requires a change to documentation (this usually means the notes in the description of a package).
- [ ] I have updated the documentation accordingly (this usually means the notes in the description of a package).
- [ ] I have updated the package description and it is less than 4000 characters.
- [ ] All files are up to date with the latest [Contributing Guidelines](https://github.com/chocolatey-community/chocolatey-packages/blob/master/CONTRIBUTING.md)
- [ ] The added/modified package passed install/uninstall in the chocolatey test environment.
- [ ] The changes only affect a single package (not including meta package).

### CODE REVIEW

File: automatic/kubernetes-cli/update.ps1 [hunk 1]
```diff
@@ -28,17 +28,22 @@ function global:au_GetLatest {
     $changelog_page = Invoke-WebRequest -Uri $changelogs -UseBasicParsing
     #Only the latest three minor versions are supported: https://github.com/kubernetes/sig-release/blob/master/release-engineering/versioning.md
     #However, there are prereleases before the version is released, so the latest four changelogs/minor versions need to be checked.
-    $minor_version_changelogs = $changelog_page.links | ? href -match "CHANGELOG-[\d\.]+\.md" | Select-Object -Expand href -First 4
-    
+    $minor_version_changelogs = $changelog_page.links `
+      | Where-Object href -match "CHANGELOG-(?<version>\d+\.\d+)\.md`$" `
+      | Select-Object -Expand href -Skip 1 -First 3
```
Comment (TheCakeIsNaOH) [inline automatic/kubernetes-cli/update.ps1:33]: This should work when there are three supported release versions, and one prerelease version, like if `1.22`, `1.23`, and `1.24` are released, and `1.25 has prerelease versions available.

But it would break once `1.25` is released if they do not have a `1.26` prerelease available.

Feel free to correct me if I am misunderstanding this.

File: automatic/kubernetes-cli/kubernetes-cli.json [hunk 2]
```diff
@@ -1,10 +1,5 @@
 ﻿{
-    "1.4":  "1.4.12",
-    "1.5":  "1.5.8",
-    "1.6":  "1.6.13",
-    "1.7":  "1.7.16",
-    "1.8":  "1.8.15",
-    "1.9":  "1.9.11",
-    "1.1":  "1.19.5",
-    "1.2":  "1.24.1"
+    "1.22":  "1.22.12",
```
Comment (TheCakeIsNaOH) [inline automatic/kubernetes-cli/kubernetes-cli.json:2]: No need to include the changes in this file, as AU will update it when it updates the package.

File: automatic/kubernetes-cli/kubernetes-cli.nuspec [hunk 3]
```diff
@@ -3,7 +3,7 @@
 <package xmlns="http://schemas.microsoft.com/packaging/2015/06/nuspec.xsd">
   <metadata>
     <id>kubernetes-cli</id>
-    <version>1.24.1</version>
```
Comment (TheCakeIsNaOH) [inline automatic/kubernetes-cli/kubernetes-cli.nuspec:6]: No need to include the changes in this file, as AU will update it when it updates the package.

File: automatic/kubernetes-cli/legal/VERIFICATION.txt [hunk 4]
```diff
@@ -4,12 +4,12 @@ Verification is intended to assist the Chocolatey moderators and community
 in verifying that this package's contents are trustworthy.
 
 The embedded software have been downloaded from the listed download
-location on <https://github.com/kubernetes/kubernetes/releases/latest>
+location on <https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.24.md>
 and can be verified by doing the following:
 
 1. Download the following:
-  32-bit software: <https://dl.k8s.io/v1.24.1/kubernetes-client-windows-386.tar.gz>
```
Comment (TheCakeIsNaOH) [inline automatic/kubernetes-cli/legal/VERIFICATION.txt:11]: No need to include the changes in this file, as AU will update it when it updates the package.

File: automatic/kubernetes-cli/update.ps1 (hunk 1 above)
Comment (rgl) [reply automatic/kubernetes-cli/update.ps1:33]: I'm not sure, but we will notice if/when this becomes a problem once 1.25/1.26 stable is (not) released.

BTW, there might be a better way for looking for the k8s versions, but I did not try to find it. I just made that part of the existing code more clear.

File: automatic/kubernetes-cli/kubernetes-cli.json (hunk 2 above)
Comment (rgl) [reply automatic/kubernetes-cli/kubernetes-cli.json]: I've dropped this files changes. Pleas verify.

File: automatic/kubernetes-cli/kubernetes-cli.nuspec (hunk 3 above)
Comment (rgl) [reply automatic/kubernetes-cli/kubernetes-cli.nuspec]: I've dropped this files changes. Pleas verify.

File: automatic/kubernetes-cli/legal/VERIFICATION.txt (hunk 4 above)
Comment (rgl) [reply automatic/kubernetes-cli/legal/VERIFICATION.txt]: I've dropped this files changes. Pleas verify.

**Review by TheCakeIsNaOH — COMMENTED** [review]
Thanks for updating this, there is still the one open question, but maybe someonelse has thoughts?

File: automatic/kubernetes-cli/update.ps1 (hunk 1 above)
Comment (rgl) [reply automatic/kubernetes-cli/update.ps1]: This is now fixed. It will download the last three released version streams.

**Comment by rgl** [comment]
> Thanks for updating this, there is still the one open question, but maybe someon else has thoughts?

It should be fixed now. Please verify.

File: automatic/kubernetes-cli/update.ps1 [hunk 5]
```diff
@@ -25,25 +25,46 @@ function global:au_SearchReplace {
...
+    $minor_versions_to_keep = 3
+
     $changelog_page = Invoke-WebRequest -Uri $changelogs -UseBasicParsing
-    #Only the latest three minor versions are supported: https://github.com/kubernetes/sig-release/blob/master/release-engineering/versioning.md
-    #However, there are prereleases before the version is released, so the latest four changelogs/minor versions need to be checked.
-    $minor_version_changelogs = $changelog_page.links | ? href -match "CHANGELOG-[\d\.]+\.md" | Select-Object -Expand href -First 4
-    
+    $minor_version_changelogs = $changelog_page.links `
+      | Where-Object href -match "CHANGELOG-(?<version>\d+\.\d+)\.md`$" `
+      | Select-Object -Expand href -First ($minor_versions_to_keep+1)
```
Comment (AdmiringWorm) [inline automatic/kubernetes-cli/update.ps1:39]: Any reason why you restrict the amount of links?

If there are no reason, this can be changed to
```suggestion
      | Select-Object -Expand href
```

So we can monitor all links (In case there are a patch being created for an older release at some point).

File: automatic/kubernetes-cli/update.ps1 [hunk 6]
```diff
@@ -25,25 +25,46 @@ function global:au_SearchReplace {
...
     $streams = @{}
-    
+
     $minor_version_changelogs | ForEach-Object {
-        $minor_version = ($_ -split "CHANGELOG-" | Select-Object -Last 1).trim(".md")
-        $minor_changelog_page = Invoke-WebRequest -UseBasicParsing -Uri ("https://github.com" + $_)
-        $url64 = $minor_changelog_page.links | ? href -match "/v(?<version>[\d\.]+)/kubernetes-client-windows-amd64.tar.gz" | Select-Object -First 1 -Expand href
+        if ($streams.Count -ge $minor_versions_to_keep) {
+          return
+        }
```
Comment (AdmiringWorm) [inline automatic/kubernetes-cli/update.ps1:46]: We shouldn't restrict the amount of streams being added.

Please remove this if statement.

File: automatic/kubernetes-cli/update.ps1 [hunk 7]
```diff
@@ -25,25 +25,46 @@ function global:au_SearchReplace {
...
+        if ($_ -notmatch "CHANGELOG-(?<version>\d+\.\d+)\.md`$") {
+          return
+        }
+        $minor_version = $matches.version
+        $minor_changelog_page = Invoke-WebRequest -UseBasicParsing -Uri "https://github.com$_"
```
Comment (AdmiringWorm) [inline automatic/kubernetes-cli/update.ps1:51]: Can we add a check before this line to see if the stream already contains the `$minor_version` so we don't need to do a web request when not needed.

**Review by AdmiringWorm — CHANGES_REQUESTED** [review]

File: automatic/kubernetes-cli/update.ps1 (hunk 5 above)
Comment (rgl) [reply automatic/kubernetes-cli/update.ps1:39]: The restrictions came from the existing package. I just made them more explicit. From what I understood, this package intent was to track only the latest version of the latest three kubernetes releases.

But now it seems you want to track all versions and releases? Even no longer supported versions? So scape all the versions that are available?

Comment (AdmiringWorm) [reply automatic/kubernetes-cli/update.ps1:39]: Typically the restriction is only in place for the beginning, to not overwhelm Chocolatey Community repository with too many versions at the same time, as well as not needing to fix a lot of package versions if they fail the verifier/validator.

Most likely, this restriction was just forgotten to be removed. However, since it already has the restriction we can leave it here for this PR.

File: automatic/kubernetes-cli/update.ps1 (hunk 6 above)
Comment (AdmiringWorm) [reply automatic/kubernetes-cli/update.ps1:46]: Although we can leave the 3 limit restriction for now, this if statement should still be removed.

File: automatic/kubernetes-cli/update.ps1 (hunk 7 above)
Comment (AdmiringWorm) [reply automatic/kubernetes-cli/update.ps1:51]: Just to note, this is more of a safe guard in case the changelog gets incorrectly updated at some point.

Comment (rgl) [reply automatic/kubernetes-cli/update.ps1:51]: Not sure you can prevent that with such a simple heuristic. For example, I guess this can go both ways. It can be flapping between wrong/right, so, I would not bother.

File: automatic/kubernetes-cli/update.ps1 (hunk 6 above)
Comment (rgl) [reply automatic/kubernetes-cli/update.ps1]: I've removed it in the last push. The output still looks like:

```json
{
  "Streams": {
    "1.22": {
      "Version": "1.22.12",
      "ReleaseNotes": "https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.22.md",
      "URL32": "https://dl.k8s.io/v1.22.12/kubernetes-client-windows-386.tar.gz",
      "URL64": "https://dl.k8s.io/v1.22.12/kubernetes-client-windows-amd64.tar.gz"
    },
    "1.23": {
      "Version": "1.23.9",
      "ReleaseNotes": "https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.23.md",
      "URL32": "https://dl.k8s.io/v1.23.9/kubernetes-client-windows-386.tar.gz",
      "URL64": "https://dl.k8s.io/v1.23.9/kubernetes-client-windows-amd64.tar.gz"
    },
    "1.24": {
      "Version": "1.24.3",
      "ReleaseNotes": "https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.24.md",
      "URL32": "https://dl.k8s.io/v1.24.3/kubernetes-client-windows-386.tar.gz",
      "URL64": "https://dl.k8s.io/v1.24.3/kubernetes-client-windows-amd64.tar.gz"
    }
  }
}
```

Should it contain all the all versions inside a given stream (although the function name is `au_GetLatest`)? Or should it only contain the latest (as its now)?

File: automatic/kubernetes-cli/update.ps1 (hunk 7 above)
Comment (AdmiringWorm) [reply automatic/kubernetes-cli/update.ps1:48]: Adding a simple if check if to see if the stream have already been added is mainly for the benefit of not aborting the entire collection.

If the changelog at some point will be incorrectly updated with the same major+minor version, then an exception will happen causing nothing to actually be opened, even the streams that was already added/will be added.

So, as I mentioned it is purely a safe guard so the updater won't throw an exception nothing else.

Basically the only thing that need to be added is:

```powershell
if ($streams.ContainsKey($minor_version)) {
  return
}
```

Comment (rgl) [reply automatic/kubernetes-cli/update.ps1:51]: Done. Please verify.

File: automatic/kubernetes-cli/update.ps1 (hunk 6 above)
Comment (AdmiringWorm) [reply automatic/kubernetes-cli/update.ps1]: Could be that something unrelated prevents additional streams from being added, I'll ignore that for now and come back to it later to check what is causing the issue.

File: automatic/kubernetes-cli/update.ps1 (hunk 7 above)
Comment (AdmiringWorm) [reply automatic/kubernetes-cli/update.ps1:51]: The changes look good, thank you for adding the `if` check.

**Review by AdmiringWorm — APPROVED** [review]
LGTM, I have updated the stream limit to 10 packages. I would normally remove the limit, but it seems we are missing quite a few versions on community.chocolatey.org which exceeds the number of versions that can be in the submitted queue.

**Comment by rgl** [comment]
Any idea why https://community.chocolatey.org/packages/kubernetes-cli still does not have the latest versions?

**Comment by AdmiringWorm** [comment]
Seems that the latest version failed to be pushed to the Community Chocolatey Repository, but AU did not see it as an error. I'll get the stream for that version reset so it will try it again.

**Comment by rgl** [comment]
It has worked. 1.24.3 is now available. Thanks for looking into it!
